### PR TITLE
[#IOPID-385]  Call notify login endpoint

### PR DESCRIPTION
**Pull request by gquadrati**
This PR depends on https://github.com/pagopa/io-spid-commons/pull/149
This PR depends on https://github.com/pagopa/io-functions-app/pull/282

#### List of Changes

- See commit history

#### Motivation and Context

Notify the user in case a new login has been performed.

#### How Has This Been Tested?

Unit test, io-mock

#### Types of changes

- [ ] Chore (nothing changes by a user perspective)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

#### Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by arcogabbo — COMMENTED** [review]
nice work, just a minor stylish comment

File: src/controllers/__tests__/authenticationController.test.ts [hunk 1]
```diff
@@ -297,7 +315,7 @@ afterEach(() => {
   clock = clock.uninstall();
 });
 
-const setupMocks = () => {
+const setupGetNewTokensMocks = () => {
```
Comment (BurnedMarshal) [inline src/controllers/__tests__/authenticationController.test.ts:318]: This change is needed? Seems to be only a rename.

File: src/controllers/__tests__/authenticationController.test.ts [hunk 2]
```diff
@@ -998,6 +1046,175 @@ describe("AuthenticationController|>LV|>acs", () => {
   );
 });
 
+describe("AuthenticationController|>LV|>acs|>notify user login", () => {
+  const setupLollipopScenario = () => {
+    jest
+      .spyOn(authCtrl, "isUserElegibleForFastLogin")
+      .mockImplementationOnce((_) => true);
```
Comment (BurnedMarshal) [inline src/controllers/__tests__/authenticationController.test.ts:1053]: We can use this instead.
```suggestion
      .mockReturnValueOnce(true);
```

File: src/controllers/authenticationController.ts [hunk 3]
```diff
@@ -214,6 +226,13 @@ export default class AuthenticationController {
             LoginTypeEnum.LEGACY,
           ];
 
+    // Retrieve user IP from request
+    const errorOrUserIp = IPString.decode(spidUser.getAcsOriginalRequest()?.ip);
+
+    if (isUserElegibleForFastLoginResult && E.isLeft(errorOrUserIp)) {
+      return ResponseErrorInternal("Error reading user IP");
+    }
```
Comment (BurnedMarshal) [inline src/controllers/authenticationController.ts:234]: We are blocking the login if, for some reason, the IP is invalid or missing. Can we consider this parameter as optional?

File: src/controllers/authenticationController.ts [hunk 4]
```diff
@@ -477,6 +499,11 @@ export default class AuthenticationController {
         // in io-spid-commons does not support 429 / 409 errors
         return ResponseErrorInternal(createProfileResponse.kind);
       }
+
+      userEmail =
+        newProfile.is_email_validated && newProfile.email
+          ? newProfile.email
+          : user.spid_email;
```
Comment (BurnedMarshal) [inline src/controllers/authenticationController.ts:506]: This is always equal to the `spid_email` or undefined for CIE logins

File: src/controllers/authenticationController.ts [hunk 5]
```diff
@@ -517,6 +550,80 @@ export default class AuthenticationController {
...
+            },
+          });
+
+          return ResponseErrorInternal("Unable to notify user login event");
+        })
+      )();
+
+      if (E.isLeft(errorOrNotifyLoginResult)) {
+        const delLollipopResult =
+          await this.sessionStorage.delLollipopDataForUser(user.fiscal_code);
```
Comment (BurnedMarshal) [inline src/controllers/authenticationController.ts:609]: What if we send the email before setting all the lollipop and session data?

File: src/controllers/__tests__/authenticationController.test.ts (hunk 1 above)
Comment (gquadrati) [reply src/controllers/__tests__/authenticationController.test.ts:318]: Yes, it's just a rename to better explain what setup is doing here

File: src/controllers/authenticationController.ts (hunk 4 above)
Comment (gquadrati) [reply src/controllers/authenticationController.ts]: Right, done in [[#IOPID-385] just take spid_email for new users](https://github.com/pagopa/io-backend/pull/1040/commits/fb9fdd352426112daef92e210381128d2e2230c2)

File: src/controllers/authenticationController.ts (hunk 5 above)
Comment (gquadrati) [reply src/controllers/authenticationController.ts]: It requires some effort, we can refactor it in another PR

File: src/controllers/authenticationController.ts (hunk 3 above)
Comment (gquadrati) [reply src/controllers/authenticationController.ts:234]: Right now IP address is required by the new notify-login endpoint.

**Review by BurnedMarshal — APPROVED** [review]

**Pull request merged**